### ISSUE TITLE
Python 3

### ISSUE THREAD

**Issue by zaufi** [nucleic/enaml#117]
any plans to make it Python 3 compatible?

**Comment by rutsky** [comment]
What is the status of the Python 3 support?

I see, that atom in 1.0.0-dev branch installs in Python 3, but also I see, that atom in 1.0.0-dev branch changed it's API and can't be used with enaml even in Python 2.

Can you estimate how hard is to port enaml and it's dependencies (atom, kiwi) to Python 3?

Is atom 1.0.0-dev changed API significantly, that requires significant rewrite of enaml?

Is porting of enaml and it's dependencies is mostly mechanic work, or it requires significant and non-obvious rewrite?

**Comment by MatthieuDartiailh** [comment]
@sccolbert I am also interested in Python 3 support. And I can participate in the porting with some guidances (I understand you are pretty busy with phosphor right now and I guess for some time).
Would porting the current version to python 3 makes sense (if somebody does the job) or have you plans for the 1.0.0 version in a not too far future?

**Comment by peterazmanov** [comment]
I made a working Python 3.4 fork of enaml [peterazmanov/enaml](https://github.com/peterazmanov/enaml). Master head commit contains changes to work with Qt5 only (tested with PyQt 5.4). This fork requires compatible atom [peterazmanov/atom](https://github.com/peterazmanov/atom) and kiwisolver ported to Python 3 [jwiggins/kiwi/python-3-support](https://github.com/jwiggins/kiwi/tree/feature/python-3-support). All changes are far from production state, require thorough testing and review. )

**Comment by peterazmanov** [comment]
@sccolbert Do you have plans to support Python 2 version of enaml in the future? I have plans to make my Python 3 fork suitable for PR, hope community will help. If support of Py2k is required it'll increase required work.

**Comment by MatthieuDartiailh** [comment]
I would find valuable to have a cross-compatible version. The future package is very nice to support both but it won't help with ast and byteplay inconsistencies.

**Comment by sccolbert** [comment]
Atom and Kiwi need to support Python 3 before I even think about Enaml support. However, I'm currently 110% occupied with PhosphorJS work.

**Comment by MatthieuDartiailh** [comment]
Even if you do not have time to actually code or review code at the moment, could you give us pointers of what remains to be discussed about let say atom-1.0 ? Discussion at Google groups stopped some months ago, and in other threads you mentioned possibilities that were quite different from the current implementation (like the possibility to remove type checking at runtime for stable application, and moving all validation to pure python). If you could just draft the points that need to be discussed, we (the community) could discuss them and come back to you with a cleanly written proposal, hence reducing the burden on your side.
As far as kiwi is concerned, @jwiggins PR seems to address that but I think we should address atom first.

**Comment by jwiggins** [comment]
:+1: I'm also interested in atom on Py3k.

**Comment by peterazmanov** [comment]
@MatthieuDartiailh let's focus on 'simple' technical task of porting to Py3k and not discuss changes in atom API. Roadmap as I see it:
1. @jwiggins [kiwi PR](https://github.com/nucleic/kiwi/pull/13) AFAICT is compatible with Py2k and can be merged
2. Decide whether Py2k support for atom required
3. Port atom to Py3k (with Py2k support if item 2. of this list is 'yes') (we can open new fork or discuss changes I made to support Py3k in atom [peterazmanov/atom](https://github.com/peterazmanov/atom))
4. Decide whether Py2k support for enaml required
5. Port enaml to Py3k (with Py2k support if item 4. of this list is 'yes') (we can open new fork or discuss changes I made to support Py3k in enaml [peterazmanov/enaml](https://github.com/peterazmanov/enaml))

**Comment by jwiggins** [comment]
@peterazmanov: Correct. The kiwi PR that I have open has been tested on both 2.7 and 3.4. I would imagine that supporting both flavors of Python in atom wouldn't be too much work. Plus, with TravisCI it's relatively straightforward to maintain support for both (as long as your test coverage is adequate).

I'm not sure how realistic simultaneous 2/3 support is for enaml. Due to the bytecode foolery, it might make for a bit of extra work.

**Comment by MatthieuDartiailh** [comment]
Just a quick update on this.
- @peterazmanov branch of atom should now be Py23 compatible.
- some works has been done on enaml too but is not complete : 
  - C code should be cross compatible
  - future is now used to smooth Py23 incompatibilities (metaclasses)
  - byteplay was reworked and both a python 2 and a python 3 version can be found here https://github.com/serprex/byteplay (but apparently it is not reasonable to merge both files together)

Next large step is to make the parser cross-compatible. I just started looking into this. The idea would be to support the same ast that the version of Python used support. I am not sure how feasible it is using ply, and I am looking at pgen2 (found in lib2to3) as a way to overcome that difficulty (I need to see if it is possible to monkeypatch the python grammar to add the enaml syntax rules). If anybody has a better idea or some pointers for me, I will accept them !

**Comment by peterazmanov** [comment]
> byteplay was reworked and both a python 2 and a python 3 version

https://github.com/serprex/byteplay contains buggy version for Python 2, see issue https://github.com/serprex/byteplay/issues/5 for details. As I mentioned in that issue version for Python 2 should be taken in its original form from [Noam Yorav-Raphael](http://byteplay.googlecode.com/svn/trunk/byteplay.py).

**Comment by MatthieuDartiailh** [comment]
I started working on the parser and finally chose to go on with ply (which is much more reasonable). I did some small scales tests and the ideas I used seem to work. This is still wip but anybody wanting a look can go there https://github.com/MatthieuDartiailh/enaml/tree/parser23 (the new files are under enaml/core/parsing).

**Comment by MatthieuDartiailh** [comment]
So far the parser works and most things works on both Python 2 and 3. I would welcome tests of Python 3 specific syntax as I have not tested all of them. Note that Python 3.5 async/await are not supported yet (I had no time to run any test on Python 3.5).
One thing is not working. Enaml does a lot of magic with scope to allow to access to widgets, define 'self' and 'change' inside notification handlers ('::'), or function call (no 'change' there). The issue comes when one uses list/dict/set comprehensions in Python. Those are written in the bytecode in term of function call and I did not find how to alter the scope (in Python 2 the problem does not exist as there is no function calls).  Any help on that would be welcome @sccolbert @peterazmanov .

**Comment by MatthieuDartiailh** [comment]
I fixed the problem by inlining the function calls in the byteplay. This has the nice side effect of allowing the use of dict and set comprehensions in Python 2 (I do not know why list were treated in a special manner). 
The inlining happens for all operators and declarative functions. It is run only if List/Set/DictComp node is present in the ast.

**Comment by sccolbert** [comment]
List comprehensions in Python 2 do not create a separate generator code
object like the others:

>>> def foo():
...     return [i for i in range[5]]
...
>>> dis.dis(foo)
  2           0 BUILD_LIST               0
              3 LOAD_GLOBAL              0 (range)
              6 LOAD_CONST               1 (5)
              9 BINARY_SUBSCR
             10 GET_ITER
        >>   11 FOR_ITER                12 (to 26)
             14 STORE_FAST               0 (i)
             17 LOAD_FAST                0 (i)
             20 LIST_APPEND              2
             23 JUMP_ABSOLUTE           11
        >>   26 RETURN_VALUE
>>> def bar():
...     return { i for i in range(5) }
...
>>> dis.dis(bar)
  2           0 LOAD_CONST               1 (<code object <setcomp> at
0251DC80, file "<stdin>", line 2>)
              3 MAKE_FUNCTION            0
              6 LOAD_GLOBAL              0 (range)
              9 LOAD_CONST               2 (5)
             12 CALL_FUNCTION            1
             15 GET_ITER
             16 CALL_FUNCTION            1
             19 RETURN_VALUE

**Comment by MatthieuDartiailh** [comment]
Ok.
Do you agree with my solution or do you have a better idea on how to run the code in the right scope ? I must say I spend quite a lot of time looking for alternative solution but it is the only one that worked.

**Comment by sccolbert** [comment]
I honestly don't have the time to take a look at it right now (sorry). I was just explaining why list comps worked when the others didn't.

**Comment by MatthieuDartiailh** [comment]
No problem. Thanks.
Perhaps once we will have polished everything, you will find time to give a look. It would be awesome to have a release with support for Python 3.

**Comment by MatthieuDartiailh** [comment]
I am quite happy to announce that with the latest version of my branch all examples run and additionally all the tests of the py3 branch of https://github.com/Ecpy/ecpy are passing (that's a lot of enaml code) run on Python 3.4.
Python 3.5 is not supported as https://github.com/jwiggins/kiwi does not support it. I will try to update it. Additionally async funcdef and associated async and await statements are not supported in the parser adding them should be possible but ensuring that async and await are not used outside of an async funcdef will require some care. I consider those as low priority.
One nice further step would be qt5 support. @sccolbert if I remember correctly when you started implementing it you created a new package and re-implented all widgets. What was the rationale behind that ? Why not simply implement the right shims inside the imports ?

**Comment by MatthieuDartiailh** [comment]
Actually another missing feature of the parser is support for keyword only arguments.

**Comment by blink1073** [comment]
:clap:  well done, @MatthieuDartiailh!

**Comment by jwiggins** [comment]
@MatthieuDartiailh: What changed in 3.5? Did the addition of the `@` operator cause unique problems for C extensions?

**Comment by MatthieuDartiailh** [comment]
I had no time to investigate actually but installing on linux via pip crashed gcc.  The one difference I know it that the tp_reserved slot is know tp_async. Feel free to have a look @jwiggins I do not think I will have time to look into it in the next few days.

Thanks @blink1073. Actually I would not have gone so far without the work of @peterazmanov on byteplay for Python 3.

**Comment by MatthieuDartiailh** [comment]
I did check and indeed numbers have too new slots. I will open a PR on your PR if it is ok with you @jwiggins.

**Comment by MatthieuDartiailh** [comment]
I fixed kiwi (https://github.com/jwiggins/kiwi/pull/3) but now enaml is segfaulting, ....

**Comment by MatthieuDartiailh** [comment]
The segfault was kind of stupid (mainly cache issues ....). However not everything is working yet on 3.5.

**Comment by MatthieuDartiailh** [comment]
My branch is now compatible with Python 3.5 (run ecpy testsuite). Furthermore I added support for keyword only arguments in functions and advanced unpacking (including dictionary unpacking for Python 3.5). As Ecpy is Python 2/3 compatible those features have not been tested very thoroughfully yet. If anybody has time to run some tests that would be great (or simply post a script testing those cases that I can use).
The last point remaining is the support for coroutines (async/await).
Out of curiosity does anybody have a copy of the qt5 branch that @sccolbert deleted some time ago ?

**Comment by sccolbert** [comment]
I ran into lots of rendering regression issues with Qt5 (bugs in Qt). I
made several patches to Qt which took ~7 months and 2 releases before they
were merged. I ran out of care at that point to continue the effort.

On Sun, Mar 6, 2016 at 2:23 PM, Matthieu Dartiailh <notifications@github.com

**Comment by MatthieuDartiailh** [comment]
I know that you had a lot of trouble with qt5. I am interested nonetheless as I know that for example on the most recent versions of OS X qt4 is not very well supported (the Spyder project had a lot of complaint about things being super slow and they appear to be fixed with qt5). I really like enaml and that's why I would like to make sure that we can go on using it by making it work on recent versions of Python and Qt.
I was wondering about your branch to avoid restarting from scratch.
I also hope to see a Phosphor backend for enaml one day !

**Comment by sccolbert** [comment]
Most of the work was just adjusting the imports to pull in Qt5. That got
~80% of the way there IIRC.

On Sun, Mar 6, 2016 at 2:52 PM, Matthieu Dartiailh <notifications@github.com

**Comment by peterazmanov** [comment]
AFAIK the amount of changes to support Qt5 is relatively small. I believe most of them are already done in my fork (not in cross-compatible fashion, sorry). See [Qt5-branch-of-my-fork](https://github.com/peterazmanov/enaml/tree/Qt5).

**Comment by blink1073** [comment]
Spyder is using this shim for Qt abstraction: https://github.com/spyder-ide/qtpy

**Comment by MatthieuDartiailh** [comment]
I added support for coroutine and fixed support for @ operator in Python 3.5 parser. Next I will try to deal with Qt4/5 cross support (in a new branch). Ithink that qtpy provides a nice abastraction so I will try to use it.
@sccolbert would accept PRs for atom/enaml/kiwi for python 3 support ? No need to say, I am willing to address any bugs that people running on Python 3 could find.

**Comment by sccolbert** [comment]
Yes. I would accept PRs, but let's do them one at a time in order of kiwi
-> atom -> enaml

On Tue, Mar 8, 2016 at 1:36 PM, Matthieu Dartiailh <notifications@github.com

**Comment by sccolbert** [comment]
And by "accept", I mean I'll try to devote some time to
review/comment/cleanup etc. The smaller each PR the better, since I can
review those faster.

**Comment by MatthieuDartiailh** [comment]
For kiwi I made a PR againt @jwiggins branch so that his PR is up to date for Python 3.5 support (https://github.com/nucleic/kiwi/pull/13). Once you are happy with that one, I will update atom and enaml based on your comment and try to open small PRs. For atom, I think we can go with only one but for enaml I will try to split them in manageable pieces.

**Comment by MatthieuDartiailh** [comment]
If you are short on time for cleaning up I will be happy to do so given some guide lines.

**Comment by jwiggins** [comment]
I ran nucleic/kiwi#13 through Travis CI with 2.7, 3.4, and 3.5 after merging the changes from @MatthieuDartiailh and doing a little bit of cleanup. It's good to go.

**Comment by sccolbert** [comment]
thanks!

**Comment by sccolbert** [comment]
Ok. The Python 3 support for kiwi has been merged. Since this was a
non-trivial change, before making a new release I'd like to get some more
tests in the repo which test all of the symbolic operators and the
correctness of the solver for some simple problems. Any volunteers for that?

**Comment by MatthieuDartiailh** [comment]
I can give it a try if nobody beats me to it. Any preference concerning the testing framework ? Personally I am a fan of pytest.

**Comment by sccolbert** [comment]
Simple is better. The current sanity tests use nose.

**Comment by MatthieuDartiailh** [comment]
Yes but it uses simply assert stmt for which nose won't provide nice tracebacks, whilefor the exact same tests pytest would provide more detailed reports. As far as the current tests are concerned we simply change the invocation.

**Comment by sccolbert** [comment]
That's fine. I'm just saying we don't need a super huge testing
infrastructure just for kiwi :)

**Comment by peterazmanov** [comment]
@MatthieuDartiailh I tested your branch on my application (~40K of enaml code with fair amount of template code) with Python 3.4.3, everything works.

**Comment by MatthieuDartiailh** [comment]
Great @peterazmanov ! thanks for testing.

**Comment by tstordyallison** [comment]
Just so I'm clear - are we thinking that we'll be able get enaml to a place where it is is cross-compat for py2 and 3? @peterazmanov I saw some super() calls etc that won't work on py2, that's all ;)

Also interested in seeing how qtpy goes... Qt5 would be nice if it 'just works' :)

**Comment by sccolbert** [comment]
This seems to be how it's shaping up.

**Comment by peterazmanov** [comment]
> I saw some super() calls etc that won't work on py2

@tstordyallison my enaml branch isn't cross-compat, @MatthieuDartiailh made significant progress in cross-compat direction (see his [branch](https://github.com/MatthieuDartiailh/enaml/tree/parser23) for details).

> Also interested in seeing how qtpy goes... Qt5 would be nice if it 'just works'

It worked for me :) I can't say my program contains some elaborate UI, but still I had to monkeypatch some desired behavior of controls (pyqt & qt problems + not all qt features are supported by enaml). Any further testing is welcome.

**Comment by MatthieuDartiailh** [comment]
My branch is fully Python 2/3 cross-compatible and support Python 2/3 only features in enaml files (for examples one can use async /await in enaml for Python 3.5).

I started some work to use qtpy to be able to use qt5. But I am waiting to get my previous work merged before going on further.

**Comment by MatthieuDartiailh** [comment]
Waiting for #208, #209, and #210 to be reviewed by Chris and merged, I uploaded conda packages for atom, kiwi and enaml on the ecpy channel. I built packages for python 2.7, 3.4 and 3.5 on Linux, OSX and windows (32 and 64) using Travis and AppVeyor. The more people will test and comment on those the better so do not hesitate to report any trouble to https://github.com/MatthieuDartiailh/atom and https://github.com/MatthieuDartiailh/enaml (for kiwi I simply built the latest master).

To install simply use : 
```shell
conda install enaml -c ecpy
```

**Comment by Tillsten** [comment]
@MatthieuDartiailh Thanks for making the py3 version conda installable, i will give it test run.

**Comment by MatthieuDartiailh** [comment]
I have one issue that I cannot figure out under python 3 it is illustrated in the following gist https://gist.github.com/MatthieuDartiailh/16cbb96a1ba142fa95ee5be947bdf1f1.

Under python 2 the traceback is correct and point to the ValueError, under python 3 it throws a KeyError saying that there is no 'aux' keyx in scope['nonlocals']. I have seen a variant (or something else I am not sure )where it is setattr that is accused of returning NULL without setting an exception.

If anybody has a brilliant idea about the origin of this discrepancy, he is welcome.

**Comment by blakejohnson** [comment]
@MatthieuDartiailh I don't have a solution to your bug, but I wanted to pipe up to say that I have seen similar problems. In my particular case, I am trying to call `remove` on a `ContainerList` and get a message like:
```
SystemError: <built-in method remove of atom.catom.atomclist object at 0x118691d00> returned NULL without setting an error
```
If I find anything, I will let you know.

**Comment by sccolbert** [comment]
@blakejohnson That error will occur when a C-API function returns `NULL` to the interpreter, but has not set the exception which should be raised. It indicates a bug in the C code.

**Comment by MatthieuDartiailh** [comment]
Could you post an example to reproduce the error ? Usually when I see that kind of error, an error (or warning see https://github.com/MatthieuDartiailh/enaml/issues/6) actually occurred deeper (during the handling of an enaml operator). So it is either an error in the C code as pointed out by @sccolbert or something harder to track in the bytecode manipulation performed by enaml.

**Comment by blakejohnson** [comment]
@sccolbert and @MatthieuDartiailh Sure, I will post the example in a new issue. I'm trying to pair it down a bit first.

**Comment by nicoddemus** [comment]
I was pointed to this thread by a friend and would like to mention that pytest has a [pytest-qt](https://github.com/pytest-dev/pytest-qt) plugin which supports PyQt5, PyQt4 and PySide.

**Comment by MatthieuDartiailh** [comment]
Closed by #227
